File: demos/stage1/python/main.py
import os
from langchain.document_loaders import TextLoader
from langchain.text_splitter import CharacterTextSplitter
from langchain.embeddings import OpenAIEmbeddings
from langchain.vectorstores import FAISS
from langchain.llms import OpenAI
from langchain.chains import RetrievalQA

def main():
    # Load document
    loader = TextLoader("document.txt")
    documents = loader.load()

    # Split text
    text_splitter = CharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
    docs = text_splitter.split_documents(documents)

    # Embeddings
    embeddings = OpenAIEmbeddings()

    # Vector store
    vectorstore = FAISS.from_documents(docs, embeddings)

    # LLM
    llm = OpenAI()

    # QA chain
    qa = RetrievalQA.from_chain_type(llm=llm, chain_type="stuff", retriever=vectorstore.as_retriever())

    # Query
    query = "What is the document about?"
    result = qa.run(query)
    print(result)

if __name__ == "__main__":
    main()

File: demos/stage1/go/main.go
package main

import (
	"fmt"
	"log"

	"github.com/tmc/langchaingo/documentloaders"
	"github.com/tmc/langchaingo/embeddings"
	"github.com/tmc/langchaingo/llms/openai"
	"github.com/tmc/langchaingo/textsplitter"
	"github.com/tmc/langchaingo/vectorstores/faiss"
)

func main() {
	// Load document
	loader := documentloaders.NewText("document.txt")
	docs, err := loader.Load()
	if err != nil {
		log.Fatal(err)
	}

	// Split text
	splitter := textsplitter.NewRecursiveCharacter(1000, 0)
	splitDocs := splitter.SplitDocuments(docs)

	// Embeddings
	emb, err := embeddings.NewOpenAI()
	if err != nil {
		log.Fatal(err)
	}

	// Vector store
	store, err := faiss.New(emb, faiss.WithIndexName("index"))
	if err != nil {
		log.Fatal(err)
	}
	err = store.AddDocuments(splitDocs)
	if err != nil {
		log.Fatal(err)
	}

	// LLM
	llm, err := openai.New()
	if err != nil {
		log.Fatal(err)
	}

	// Retrieval QA (simplified)
	query := "What is the document about?"
	results, err := store.SimilaritySearch(query, 5)
	if err != nil {
		log.Fatal(err)
	}

	context := ""
	for _, doc := range results {
		context += doc.PageContent + "\n"
	}

	prompt := fmt.Sprintf("Based on the following context, answer the question: %s\n\nContext: %s", query, context)
	answer, err := llm.GenerateContent(prompt)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println(answer)
}